Reduce intrusive_ptr incref/decref costs

qizzzh opened the pull request:
Summary:
Intrusive_ptr doesn't provide a explicit incref method. When a users want to
incref the target, they creates a intrusive_ptr to wrap the target, then makes
a copy which does the actual incref, then release both the first intrusive_ptr
and the copy to prevent decref at deconstruction time. This is very
inefficient. Instead, do the incref/decref directly.

Differential Revision: D18798505

smessmer commented on c10/util/intrusive_ptr.h:
```diff
@@ -686,23 +695,25 @@ using weak_intrusive_ptr_target = intrusive_ptr_target;
-    c10::intrusive_ptr<intrusive_ptr_target>::reclaim(self);
-    // NB: Caller still has 'self' pointer, but it's now invalid.
-    // If you want more safety, used the actual c10::intrusive_ptr class
+    if (self && --self->refcount_ == 0) {
```
The reason for `incref`/`decref` not actually doing incref/decref is that any logic touching the refcounter is complex, easy to get wrong (especially how refcount and weakcount interact with each other) and should be constrained to as small a surface as possible.

This is why only the `intrusive_ptr` class itself was allowed to touch `refcount_` and `weakcount_` and anything outside (i.e. even the `incref`/`decref` functions) were not allowed to touch it.

I agree though that this made `incref()` a bit more complicated than it would have been otherwise. So I'm ok with changing `incref()` if you think this is better.

For `decref()`, however this change makes things strictly worse. `decref` is now more complicated than it was before, and it is basically a duplication of `intrusive_ptr::reset()`. Can we avoid the duplication somehow? If not, let's just keep it as it was. See also my general comment on why these changes don't change perf and it's probably side-stepping the assertion that causes the perf improvements. So if we use a different way to side-step the assertion, then these changes here aren't needed anymore.

qizzzh replied:
You are providing a incref that does incref but in a different way, it's no better than just increment refcount_ inside incref. And you are also relying on compiler to optimize unnecessary code away, which is not reliable and in our case indeed it doesn't do what you expected. You are relying on intrusive_ptr destructor to decref right? Why is this worse?

smessmer reviewed (changes requested):
I tested on godbolt with gcc 5, gcc 9 and clang 7. All with '-O2'. I confirmed that if the assertion in intrusive_ptr::reclaim() is commented out, then the changes to incref() here produce exactly the same assembly as without these changes. Thanks to inlining. I didn't find a compiler that produces different assembly.

With this in mind, it seems that disabling or optimizing the assertion is the important step for perf here. The changes here accidentally improve perf because they sidestep the assertion in `reclaim()`. If this holds true and the assertion in `reclaim()` is the real issue, then I'd rather just remove that assertion and keep the rest of the code as is. This might even give us mor˜e perf gain than this diff does, because `reclaim()` is also called from other code, not only from `incref/decref`.

Also, if we can figure out a way to keep the assertion and still get the perf benefit, then that'd be preferable. This assertion has a reason for existing and did catch a bug in the past. For example, if the assertion is not expensive because of the branching but because of the atomic access inside of its condition, then that can probably be fixed because incref() needs to access and write to that atomic anyways. The assertion check is for the same atomic and the whole thing could be combined into one atomic access.

qizzzh commented:
@smessmer Can you paste your results and what you tried?  I added some simple code
`  class Class final : public intrusive_ptr_target {};

  void myfunc(c10::intrusive_ptr<Class>& a) {
      raw::intrusive_ptr::incref(a.get());
  }`
to intrusive_ptr.h and the generated assembly for myfunc shows things didn't get optimized away.

yinghai commented:
> I tested on godbolt with gcc 5, gcc 9 and clang 7. All with '-O2'. I confirmed that if the assertion in intrusive_ptr::reclaim() is commented out, then the changes to incref() here produce exactly the same assembly as without these changes. Thanks to inlining. I didn't find a compiler that produces different assembly.

Please, when you talk about something, paste the data.

qizzzh commented:
Also the assert is not a very useful one. If a pointer is invalid the refcount_ field can be anything. This assert just hopes that the invalid refcount_ <= 0 which is not guaranteed. And it's totally possible that the invalid refcount_ is > 0 and later it fails somewhere and this assert will confuse people since it looks like it's catching invalid pointers though it may not.

smessmer commented:
Sorry, I had sent the godbolt link to @qizzzh some days ago but forgot to post it here. Here we go: https://godbolt.org/z/PFZjha . Take a look at the assembly for `func()`.

qizzzh commented:
@smessmer The points are 1. Better clarity. 2. As said some compiler optimizations may not work reliably, and if we don't have to do things in a fragile way we should just avoid it.

smessmer commented on c10/util/intrusive_ptr.h:
```diff
@@ -686,23 +687,14 @@ using weak_intrusive_ptr_target = intrusive_ptr_target;
   inline void decref(intrusive_ptr_target* self) {
-    // Let it die
     c10::intrusive_ptr<intrusive_ptr_target>::reclaim(self);
-    // NB: Caller still has 'self' pointer, but it's now invalid.
```
Let's not remove this comment

smessmer commented on c10/util/intrusive_ptr.h:
```diff
@@ -730,10 +722,7 @@ namespace weak_intrusive_ptr {
   inline void decref(weak_intrusive_ptr_target* self) {
-    // Let it die
     c10::weak_intrusive_ptr<intrusive_ptr_target>::reclaim(self);
-    // NB: You still "have" the 'self' pointer, but it's now invalid.
```
neither this one

smessmer commented on c10/util/intrusive_ptr.h:
```diff
@@ -686,23 +687,14 @@ using weak_intrusive_ptr_target = intrusive_ptr_target;
+    }
   }
 
-  // WARNING: Unlike the reclaim() API, it is NOT valid to pass
```
Is it now valid to pass in `NullType::singleton`? Otherwise maybe we should keep this comment

smessmer reviewed (approved):
@qizzzh I saw you removed the changes to `decref()` and removed the assertion from `reclaim()`. This fixes my main concerns. Did you test if this removal of the assertion gave you even more performance improvement than your original approach? Places calling `reclaim()` without incref should be faster too now and I'd be interested in knowing how often we do that, i.e. if that is a critical path.

About the additional refactoring you did to `incref()` on top of the assertion removal, i.e. accessing the refcounter directly, I'm on the fence. The new implementation is shorter and easier to read, but only as long as you don't know that there's a `refcount_` and `weakcount_` and they're two different things. I don't think somebody new to this code would be able to immediately answer questions like "Shouldn't `weakcount` also be incremented in `incref()`?" so that makes it error prone to access the refcounter directly outside of the intrusive_ptr class. But this is not as strong a concern as I had for the other two points. I added a few more comments about comments that shouldn't be removed, but apart from that I'm fine with landing this as-is.

Commit:
Reduce intrusive_ptr incref/decref costs (#30709)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/30709

Intrusive_ptr doesn't provide a explicit incref method. When a users want to
incref the target, they creates a intrusive_ptr to wrap the target, then makes
a copy which does the actual incref, then release both the first intrusive_ptr
and the copy to prevent decref at deconstruction time. This is very
inefficient. Instead, do the incref/decref directly.

Differential Revision: D18798505

fbshipit-source-id: 77e6847fb37ce698d0133cc1713483e0082700a6

ezyang commented:
>  I don't think somebody new to this code would be able to immediately answer questions like "Shouldn't weakcount also be incremented in incref()?"

On the other hand, code like `intrusive_ptr` basically should *never* need changes, except bug fixes like this. So I'm not sure I would spend very much time trying to make the "onboarding" experience for the code nice.

@smessmer Are there other asserts in `intrusive_ptr` that we should watch out for?

smessmer commented:
> > I don't think somebody new to this code would be able to immediately answer questions like "Shouldn't weakcount also be incremented in incref()?"
> 
> On the other hand, code like `intrusive_ptr` basically should _never_ need changes, except bug fixes like this. So I'm not sure I would spend very much time trying to make the "onboarding" experience for the code nice.

True, but then we can just keep the code as is and not take the changes from this PR. As said before, this PR does two things: (1) improve perf and (2) do additional refactoring changes that are probably unnecessary for perf. I don't disagree with (1), only with (2). But as I said, the current state is something I can live with being landed, @qizzzh seems to have a strong opinion about it and the things I had a strong opinion about are fixed.

> @smessmer Are there other asserts in intrusive_ptr that we should watch out for?

Yes, there's more assertions. We should probably decide on a general path about how we want to use assertions. Do we want to have a rule "no assertions in hot paths" and delete them all? This would make it more likely to introduce bugs later on. Do we want to have a build mode that builds with assertions disabled for scenarios where perf matters? This is probably better than just deleting the assertions, but it introduces a difference between debug/release builds.

ezyang commented:
@smessmer Our assertions are a bit weird: we never get rid of them, even in release builds. The idea behind this was originally, "We'll look into it if they make things slow." They're making things slow. The very easiest thing you can do is `#ifdef DEBUG` these asserts to make them actually go away in release builds. Being afraid of debug/release build is a weird thing to be afraid of: it's standard practice to not put your asserts in release!

jonmorton commented:
> Do we want to have a build mode that builds with assertions disabled for scenarios where perf matters? This is probably better than just deleting the assertions, but it introduces a difference between debug/release builds.

@smessmer @ezyang this really boggles the mind, that this is even a question in a framework which claims to be production ready! It's a standard software engineering practice, in hundreds of mature, stable, performant projects, to use asserts that are disabled in debug builds...
